polus-file-renaming-plugin/src/main.py

```diff
@@ -0,0 +1,297 @@
+        rgx_pattern = re.sub(fr"\{{{named_grp}:.*?\}}", regex_str, rgx_pattern)
+    logger.debug("pattern_to_raw_f_string() returns {}.".format(rgx_pattern))
+    return rgx_pattern
+
+def pattern_to_fstring(out_pattern:str)->str:
+    """
+    Convert outpattern to format string, :03d. 
+    
+    For example, "newdata_x{row:ddd}" returns "new_data{row:03d}".
+  
+    Args:
+        out_pattern: output file pattern provided by the user
+    
+    Returns:
+        out_pattern_fstring: output file pattern converted to f-string
+    """
+    #: Return reference to logger instance
+    logger = logging.getLogger("main")
+    logger.debug("pattern_to_fstring() inputs: {}".format(out_pattern))
+    #: Make list of (named groups, user file pattern) tuples 
+    out_patterns = re.findall(r"\{(\w+):([dc+]+)\}", out_pattern)
+    f_string_dict = {}
+    for key, value in out_patterns:
+        temp_value = value[:1]
+        if "+" not in value and temp_value == "c":
+            temp_value = "s"
+            f_string_dict[key] = "{" + key + ":" + str(len(value)) + temp_value + "}"
+        # Prepend "0" to field to enable 0-padding of numeric types
+        elif "+" not in value and temp_value != "c":
+            f_string_dict[key] = "{" + key + ":0" + str(len(value)) + temp_value + "}"
+        elif "+" in value and temp_value == "c":
+            temp_value = "s"
+            f_string_dict[key] = "{" + key  + ":" + temp_value + "}"
+        # Prepend "0" to field to enable 0-padding of numeric types
+        elif "+" in value and temp_value != "c":
+            f_string_dict[key] = "{" + key  + ":0" + temp_value + "}"
+    out_pattern_fstring = out_pattern
+    for named_group, fstring in f_string_dict.items():
+        out_pattern_fstring = re.sub(fr"\{{{named_group}:.*?\}}", fstring, out_pattern_fstring)
+    logger.debug("pattern_to_f_string() returns {}.".format(out_pattern_fstring))
+    return out_pattern_fstring
+
+def replace_cat_label(inp_pattern:str, out_pattern:str)->list:
+    """Return categorical labels if inpatt is char and outpatt is digit.
+    
+    Args:
+        inp_pattern: user input pattern
+        out_pattern: user output pattern
+    Returns:
+        unique_keys: list of unique category labels to make numeric
+    """
+    #: Return reference to logger instance
+    logger = logging.getLogger("main")
+    logger.debug(
+        "replace_cat_label() inputs: {}, {}".format(inp_pattern, out_pattern))
+    #: Generate list [("row", "dd"), ("col", "dd"), ("channel", "c+"")]
+    in_patts = re.findall(r"\{(\w+):([dc+]+)\}", inp_pattern)
+    out_patts = re.findall(r"\{(\w+):([dc+]+)\}", out_pattern)
+    unique_keys = []
+    #: If input file pattern is c and output is d, list unique key
+    unique_keys = list(set([
+        inp_grp for (inp_grp,inp_rgx) in in_patts 
+        for (out_grp, out_rgx) in out_patts  
+        if inp_rgx.startswith("c") and out_rgx.startswith("d")
+        ]))
```

Comment:
Is this assuming there is only one categorical variable we want to convert to numeric? What happens if we have two string variables we want to convert to numeric variables?

Reply:
It can handle multiple categorical variables to convert to numeric. I added new tests to show it processing two and three string variables we want to convert to numeric (`test_two_chan_to_digit`, `test_three_chan_to_digit`)

If we have two string variables to convert to numeric, `non_numstr_value_to_int()` produces a dictionary for each named group where the key is the captured string for the named group and the value is the index of the string's alphabetical position. For example, `non_numstr_value_to_int()` would return `{'DAPI': 0, 'GFP': 1, 'TXRED': 2}` for one capture group (ex: `{chan:c+}`), then, `non_numstr_value_to_int()` would return `{'ABC': 0, 'DEF': 1, 'GHI': 2}` for the next capture group (ex: `{alpha_chan:ccc}`). 

Then, these values get processed and incorporated into the appropriate positions when generating the output filename.